dali/pipeline/data/tensor_vector.cc

```diff
@@ -464,107 +608,81 @@ void TensorVector<Backend>::set_type(DALIDataType new_type_id) {
                            type_.id(), "' trying to set: '", new_type_id,
                            "'. You may change the current type using Resize or by"
                            " calling Reset first."));
+  contiguous_buffer_.set_type(new_type_id);
   type_ = TypeTable::GetTypeInfo(new_type_id);
-  tl_->set_type(new_type_id);
-  for (auto t : tensors_) {
-    t->set_type(new_type_id);
-  }
-  if (state_ == State::contiguous) {
-    UpdateViews();
+  if (state_.IsContiguous()) {
+    recreate_views();
+  } else {
+    for (auto &t : tensors_) {
+      t.set_type(new_type_id);
+    }
   }
 }
 
 
 template <typename Backend>
 DALIDataType TensorVector<Backend>::type() const {
-  if (state_ == State::contiguous) {
-    return tl_->type();
-  }
-  if (curr_num_tensors_ == 0) {
-    return type_.id();
-  }
-  for (int i = 1; i < curr_num_tensors_; i++) {
-    assert(tensors_[0]->type() == tensors_[i]->type());
-  }
-  return tensors_[0]->type();
+  return type_.id();
 }
 
+
 template <typename Backend>
 const TypeInfo &TensorVector<Backend>::type_info() const {
-  if (state_ == State::contiguous) {
-    return tl_->type_info();
-  }
-  if (curr_num_tensors_ == 0) {
-    return type_;
-  }
-  for (int i = 1; i < curr_num_tensors_; i++) {
-    assert(tensors_[0]->type() == tensors_[i]->type());
-  }
-  return tensors_[0]->type_info();
+  return type_;
 }
 
 
 template <typename Backend>
 void TensorVector<Backend>::SetLayout(const TensorLayout &layout) {
-  if (state_ == State::noncontiguous) {
-    DALI_ENFORCE(!tensors_.empty(), "Layout cannot be set uniformly for empty batch");
-  }
-  tl_->SetLayout(layout);
-  for (auto t : tensors_) {
-    t->SetLayout(layout);
+  for (auto &t : tensors_) {
+    t.SetLayout(layout);
```

Comment:
I don't think it would make much sense. There is one uniform `layout_`, the `tensors_` is kinda legacy part, that is still used in SampleWorkspace and mostly it. I think we could change them to plain buffers and drop the shape, layout etc this way, and we won't lose any information.